docs/book/v3/migration/refactoring-view-helpers.md

```diff
@@ -0,0 +1,135 @@
+    }
+}
+```
+
+There are a number of problems with the class above that relied on hidden knowledge of how `AbstractHelper` worked and in turn how the plugin manager automatically injected the primary view renderer into all helpers, and also that the thing labelled 'url' or 'escapeHtml' was actually a specific type of plugin _(That may also be unexpectedly aliased to a different type)_.
+
+That's a _lot_ of hidden complexity and indirection, and furthermore it is very difficult to test.
+
+With the removal of automatic, runtime view renderer injection, and the removal of `AbstractHelper`, you are now forced to practice dependency injection in view helpers.
+
+Whilst this is inconvenient now, it will pay future dividends with more maintainable, testable and readable code.
+
+To refactor the above plugin, we would now write the following to achieve the same outcome:
+
+```php
+use Laminas\Escaper\EscaperInterface;
+use Laminas\View\Helper\HelperInterface;
+use Mezzio\Helper\UrlHelperInterface;
+
+final readonly class PluginWithDependencies implements HelperInterface
+{
+    public function __construct(
+        private UrlHelperInterface $urlHelper,
+        private EscaperInterface $escaper,
+    ) {
+    }
+    
+    public function __invoke(string $routeName, string $text): string
+    {
+        return sprintf(
+            '<a href="%s">%s</a>',
+            $this->urlHelper->generate($routeName),
+            $this->escaper->escapeHtml($text),
+        );
+    }
+}
+```
+
+The above will need a factory too:
+
+```php
+use Laminas\Escaper\EscaperInterface;
+use Laminas\View\HelperPluginManagerInterface;
+use Mezzio\Helper\UrlHelperInterface;
+use Psr\Container\ContainerInterface;
+
+final readonly class PluginFactory
+{    
+    public function __invoke(ContainerInterface $container): PluginWithDependencies
+    {
+        $helpers = $container->get(HelperPluginManagerInterface::class);
+        
+        return new PluginWithDependencies(
+            $helpers->get(UrlHelperInterface::class),
+            $container->get(EscaperInterface::class),
+        );
+    }
+}
+```
+
+You will already be registering custom plugins in configuration, but this configuration will need updating to reference the new factory:
+
+```php
+return [
+    'view_helpers' => [
+        'factories' => [
+            PluginWithDependencies::class => PluginFactory::class,
+        ],
+        'aliases' => [
+            'myPlugin' => PluginWithDependencies::class,
+        ],
+    ],
+];
+```
+
+> NOTE:
+> The key defined under `aliases` is how you _access_ the helper from within view scripts.
+> See the [Advanced Usage chapter](../helpers/advanced-usage.md#writing--registering-custom-helpers) for further information.
+
+The refactored plugin is actually compatible with both `laminas-view` version 2 _and_ version 3.
+It has always been possible to use dependency injection with `laminas-view`, therefore refactoring of plugins can happen prior to upgrading.
```

Comment:
Great detail and I think a link from the chapter "[Preparing for Version 3](https://docs.laminas.dev/laminas-view/v2/migration/preparing-for-v3/)" in documentation of version 2 would be good.